cmd/waku/server/rest/filter.go

```diff
@@ -0,0 +1,265 @@
+// NOTE: subscribe on filter client randomly selects a peer if missing for given pubSubTopic
+func (s *FilterService) subscribe(w http.ResponseWriter, req *http.Request) {
+	message := filterSubscriptionRequest{}
+	if !s.readBody(w, req, &message) {
+		return
+	}
+
+	//
+	_, err := s.node.FilterLightnode().Subscribe(req.Context(), toProtocolContentFilter(message),
+		filter.WithRequestID(message.RequestId))
+	if err != nil {
+		s.log.Error("subscription failed", zap.Error(err))
+		writeResponse(w, filterSubscriptionResponse{
+			RequestId:  message.RequestId,
+			StatusDesc: "subscription failed",
+		}, http.StatusServiceUnavailable)
+		return
+	}
+
+	// on success
+	writeResponse(w, filterSubscriptionResponse{
+		RequestId:  message.RequestId,
+		StatusDesc: http.StatusText(http.StatusOK),
+	}, http.StatusOK)
+}
+
+// 400 on invalid request
+// 500 on failed subscription
+// 200 on successful unsubscribe
+// NOTE: unsubscribe on filter client will remove subscription from all peers with matching pubSubTopic, if peerId is not provided
+// to match functionality in nwaku, we will randomly select a peer that supports filter protocol.
+func (s *FilterService) unsubscribe(w http.ResponseWriter, req *http.Request) {
+	message := filterSubscriptionRequest{} // as pubSubTopics can also be present
+	if !s.readBody(w, req, &message) {
+		return
+	}
+
+	peerId := s.getRandomFilterPeer(req.Context(), message.RequestId, w)
+	if peerId == "" {
+		return
+	}
+
+	// unsubscribe on filter
+	_, err := s.node.FilterLightnode().Unsubscribe(
+		req.Context(),
+		toProtocolContentFilter(message),
+		filter.WithRequestID(message.RequestId),
+		filter.WithPeer(peerId),
+	)
+	if err != nil {
+		s.log.Error("unsubscribe failed", zap.Error(err))
+		writeResponse(w, filterSubscriptionResponse{
+			RequestId:  message.RequestId,
+			StatusDesc: err.Error(),
+		}, http.StatusServiceUnavailable)
+		return
+	}
+
+	// on success
+	writeResponse(w, filterSubscriptionResponse{
+		RequestId:  message.RequestId,
+		StatusDesc: http.StatusText(http.StatusOK),
+	}, http.StatusOK)
+}
+
+// ///////////////////////
+// ///////////////////////
+type filterUnsubscribeAllRequest struct {
+	RequestId filterRequestId `json:"requestId"`
+}
+
+func (s *FilterService) readBody(w http.ResponseWriter, req *http.Request, message interface{}) bool {
+	decoder := json.NewDecoder(req.Body)
+	if err := decoder.Decode(message); err != nil {
+		s.log.Error("bad request", zap.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return false
+	}
+	defer req.Body.Close()
+	return true
+}
+
+// 400 on invalid request
+// 500 on failed subscription
+// 200 on all successful unsubscribe
+// unsubscribe all subscriptions for a given peer
+func (s *FilterService) unsubscribeAll(w http.ResponseWriter, req *http.Request) {
+	message := filterUnsubscribeAllRequest{}
+	if !s.readBody(w, req, &message) {
+		return
+	}
+
+	peerId := s.getRandomFilterPeer(req.Context(), message.RequestId, w)
```

Comment:
UnsubscribeAll doesn't need a peerID to be passed, as it unsubscribes from all existing filter subscriptions towards all peers. 

See https://github.com/waku-org/go-waku/blob/2701a38b2aef3b1ece7b2aef7c8df13a88e069c5/waku/v2/protocol/filter/client.go#L571

Reply:
Ok, got it.